### DISCUSSION TITLE
How do you extend a model?

### DISCUSSION THREAD

**Discussion by MichaelGradek** [dynamoose/dynamoose#1313] [Q&A]
I want to group all of my business logic for a specific entity inside a class. Following the Cat example:

```typescript
import * as dynamoose from "dynamoose";
import {Document} from "dynamoose/dist/Document";

// Strongly typed model
class Cat extends Document {
    id: number;
    name: string;
}
const CatModel = dynamoose.model<Cat>("Cat", {"id": Number, "name": String});

// Will raise type checking error as random is not a valid field.
CatModel.create({"id": 1, "random": "string"});

// Will return the correct type of Cat
const cat = await CatModel.get(1);
```

If I want to enable a Cat to meow, I'd typically change the Cat class that extends Document to something like:

```typescript
class Cat extends Document {
    id: number;
    name: string;

    meow() {
        console.log('Meow!')
    }
}
```

However, when I do that, the `meow` method is not attached to the Cat class.

How would you approach this "domain" style design using dynamoose?

**Comment by fishcharlie** [comment]
Take a look at these two methods.

- model.methods.set (https://dynamoosejs.com/guide/Model#modelmethodssetname-function)
- model.methods.document.set (https://dynamoosejs.com/guide/Model#modelmethodsdocumentsetname-function)

Would be cool someday to see if we could integrate that into the class system, but I can't think off the top of my head how that would be possible.

**Comment by andrewda** [comment]
This is something I've been thinking about for awhile, as well. Something like [`class-transformer`](https://github.com/typestack/class-transformer#plaintoclass) might help, but I'd also like to look more deeply into how other ORMs handle this.

**Reply by fishcharlie** [reply]
We should also maybe look at typegoose at some point to see how they do some of this.

**Reply by MichaelGradek** [reply]
I find the way the official AWS Data Mapper for DynamoDB solves this to be fairly intuitive:

https://github.com/awslabs/dynamodb-data-mapper-js

With TypeScript it's even more intuitive than JS.

In JS, what I do is define the Domain class on one side, and the schema object separately, and then just join them like so:

```javascript
// Cat class omitted for brevity, but contains all methods for handling the Cat domain business logic
// CatSchema omitted for brevity

Object.defineProperties(Cat.prototype, {
  [DynamoDbTable]: {
    value: 'cats',
  },
  [DynamoDbSchema]: {
    value: CatSchema,
  },
})
```
